src/item.cpp

```diff
@@ -6395,174 +6397,142 @@ void item::update_temp( const int temp, const float insulation )
-            item_counter -= loop_diff;
-            if( item_counter < 0 ) {
-                item_counter = 0;
-                return;
+            freeze_percentage = 0;
+            true_energy = completely_frozen_energy - mass * specific_heat_solid * ( freezing_temperature - env_temperature );
+        }
+    } else if( energy_change > 0 && new_item_temperature > env_temperature ) {
+        new_item_temperature = env_temperature;
+        if( env_temperature <= freezing_temperature ) {
+            true_energy = completely_frozen_energy - mass * specific_heat_solid * ( freezing_temperature - env_temperature );
+            freeze_percentage = 1;
+        } else {
+            freeze_percentage = 0;
+            true_energy = completely_liquid_energy + mass * specific_heat_liquid * ( env_temperature - freezing_temperature );
+        } 
+    }
+    if( item_tags.count( "FROZEN" ) ) {
+        item_tags.erase( "FROZEN" );
+        if ( freeze_percentage > 0.5 ) {
+            // Item melts and becomes mushy
+            apply_freezerburn();
+        }
+    } else if( item_tags.count( "COLD" ) ) {
+        item_tags.erase( "COLD" );
+    } else if( item_tags.count( "WARM" ) ) {
+        item_tags.erase( "WARM" );
+    } else if( item_tags.count( "HOT" ) ) {
+        item_tags.erase( "HOT" );
+    }
+    // Temperature tags
+    // Hot = over 50 C (323.15 K)
+    // Warm = over 30 C (303.15 K)
+    // Cold = below 5 C (278.15 K)
+    // Frozen = Over 50% frozen
+    if( new_item_temperature > 323.15 ) {
+        item_tags.insert( "HOT" );
+    } else if( new_item_temperature > 303.15 ) {
+        item_tags.insert( "WARM" );
+    } else if( freeze_percentage > 0.5 ) {
+        item_tags.insert( "FROZEN" );
+        // If below freezing temp AND the food may have parasites AND food does not have "NO_PARASITES" tag then add the "NO_PARASITES" tag.
+        if( new_item_temperature < freezing_temperature && type->comestible->parasites > 0) {
+            if( !( item_tags.count( "NO_PARASITES" ) ) ) {
+                item_tags.insert( "NO_PARASITES" );
             }
         }
+    } else if( new_item_temperature < 268.15 ) {
+        item_tags.insert( "COLD" );
+    }
+    
+    // There are extra 0.5 in here for rounding
+    true_energy = 2 * true_energy + 0.5;
+    
+    if( true_energy > INT_MAX ) {
+        debugmsg( "Item energy overflowed. The item has mass of %f", mass );
+    }
+    temperature = static_cast<int>( 10 * new_item_temperature + 0.5 ); //The extra 0.5 is there to make rounding go better
+    thermal_energy = static_cast<int>( true_energy );
+}
 
-        loop_diff = std::min( 600, diff );
-        if( diff >= 600 ) {
-            diff -= 600;
-        } else  {
-            diff = 0;
-        }
-    } while( loop_diff > 0 || item_counter < 0 || item_counter > 600 );
+float item::get_item_thermal_energy()
+{
+    float true_energy = 0.5 * thermal_energy;
+    return true_energy;
```

Comment:
You could merge two lines above into one without declaring new variable:

`return 0.5 * thermal_energy;`